### ISSUE TITLE
regarding understanding the contrastive loss code

### ISSUE THREAD

**Issue by seyeeet** [wvangansbeke/Unsupervised-Semantic-Segmentation#13]
thanks for the great work.
I am new to this topic and I am trying to understand part of the code and really appreciate it if you can explain it here to clarify it for me.
[here ](https://github.com/wvangansbeke/Unsupervised-Semantic-Segmentation/blob/aa3d0bf1dda16b7b01feae948853fbb7333ebc1d/pretrain/modules/moco/builder.py#L136)in the code I see
``` python
offset = torch.arange(0, 2 * batch_size, 2).to(sal_q.device)
sal_q = (sal_q + torch.reshape(offset, [-1, 1, 1]))*sal_q # all bg's to 0
sal_q = sal_q.view(-1)
mask_indexes = torch.nonzero((sal_q)).view(-1).squeeze()
sal_q = torch.index_select(sal_q, index=mask_indexes, dim=0) // 2
```
but I was not able to get the point here. like what is the goal of `offset ` and  what `mask_indexes ` and `sal_q` will do at the end?

Thank you very much in advance for your help.

**Comment by seyeeet** [comment]
I got it! very smart trick to make the areas that have value 1 to be come the same as the batch number that they are in!

**Comment by seyeeet** [comment]
@wvangansbeke 

if I understand it correctly when you compute the [contrastive loss](https://github.com/wvangansbeke/Unsupervised-Semantic-Segmentation/blob/9f01c78a59ab4af369734e13d65da55c48568de4/pretrain/utils/train_utils.py#L39) you are using some kind of a [labels that their values are the same as batch they are in](https://github.com/wvangansbeke/Unsupervised-Semantic-Segmentation/blob/aa3d0bf1dda16b7b01feae948853fbb7333ebc1d/pretrain/modules/moco/builder.py#L136) and then use CE.
q1) in normal contrastive loss the loss is forced to become zero but here we are forcing it to be the same as the labels that we gave to `sal_q `.
can you please help me understand the intuition? what if we have images from the same category in different batches, wouldnot that cause an issue? in general your feed back would be really appreciated :)

q2) just a follow up
```
logits.shape: torch.Size([107240, 264])
labels.shape: torch.Size([107240])
```
wouldnot the fact that we have `dim 1 = 264` cause any issue here? I am not getting the intuition here

**Comment by wvangansbeke** [comment]
If this is still an issue, better open a new one.
